gh-142048: Fix quadratically increasing GC delays

kevmo314 opened the pull request:
Fixes quadratically increasing GC delays in the free-threaded build. This is done by updating the local to global allocation count propagation to require the global count to be non-negative. This matches the existing behavior in non-free-threaded `gc.c`. Additionally, adds a flush that I found was causing allocations to be lost.

<!-- gh-issue-number: gh-142048 -->
* Issue: gh-142048
<!-- /gh-issue-number -->

kevmo314 committed:
reset allocation counts in gc thread-local buffers

kevmo314 committed:
Merge branch 'main' into fix-gc-alloc-count-negative

kevmo314 committed:
Fix allocation count decrement and GC state update

Adjust allocation count handling to prevent negative values and update GC state accordingly.

kevmo314 committed:
Fix formatting in gc_free_threading.c

colesbury commented on Python/pystate.c:
```diff
@@ -1810,6 +1810,14 @@ tstate_delete_common(PyThreadState *tstate, int release_gil)
+    // before the thread state is deleted, otherwise the count is lost.
+    _Py_atomic_add_int(&tstate->interp->gc.young.count,
+                       (int)((_PyThreadStateImpl *)tstate)->gc.alloc_count);
+    ((_PyThreadStateImpl *)tstate)->gc.alloc_count = 0;
```
```suggestion
    _Py_atomic_add_int(&tstate->interp->gc.young.count,
                       (int)tstate_impl->gc.alloc_count);
    tstate_impl->gc.alloc_count = 0;
```

kevmo314 committed:
Update Python/pystate.c

Co-authored-by: Sam Gross <colesbury@gmail.com>

kevmo314 committed:
news and lint

colesbury commented on Python/pystate.c:
```diff
@@ -1810,6 +1810,14 @@ tstate_delete_common(PyThreadState *tstate, int release_gil)
+                       (int)tstate_impl->gc.alloc_count);
+    tstate_impl->gc.alloc_count = 0;
+#endif
+
```
Let's remove this for now. It's not essential for the fix and we can add it back correctly in a subsequent PR.

This is causing a thread sanitizer failure. You can see it in the [job](https://github.com/python/cpython/actions/runs/19836858095/job/56836147631?pr=142051) if you expand the "Display logs" section:

```
WARNING: ThreadSanitizer: data race (pid=19505)
  Atomic write of size 4 at 0x55bb91207c44 by thread T52:
    #0 _Py_atomic_add_int /home/runner/work/cpython/cpython/./Include/cpython/pyatomic_gcc.h:15:10 (python+0x63520e) (BuildId: c37d32af53b8782ca7323d5cb2febaf42023f6ab)
    #1 tstate_delete_common /home/runner/work/cpython/cpython/Python/pystate.c:1816:5 (python+0x63520e)
    #2 _PyThreadState_DeleteCurrent /home/runner/work/cpython/cpython/Python/pystate.c:1893:5 (python+0x635582) (BuildId: c37d32af53b8782ca7323d5cb2febaf42023f6ab)
...

  Previous write of size 4 at 0x55bb91207c44 by main thread:
    #0 gc_collect_internal /home/runner/work/cpython/cpython/Python/gc_free_threading.c:2240:33 (python+0x5940bb) (BuildId: c37d32af53b8782ca7323d5cb2febaf42023f6ab)
    #1 gc_collect_main /home/runner/work/cpython/cpython/Python/gc_free_threading.c:2414:5 (python+0x59199c) (BuildId: c37d32af53b8782ca7323d5cb2febaf42023f6ab)
    #2 _PyGC_Collect /home/runner/work/cpython/cpython/Python/gc_free_threading.c:2723:12 (python+0x591cb7) (BuildId: c37d32af53b8782ca7323d5cb2febaf42023f6ab)
...
```

The problem is that `gc_collect_internal` reads this during a stop the world pause, but this thread is no longer considered active (i.e., we have a "detached" thread state that's getting deleted). 

We could move this to `PyThreadState_Clear()`, but let's just not include it for now. I'd like to keep the critical fix small for backporting and "losing" a few local counts probably doesn't matter.

kevmo314 replied:
Makes sense, done!

kevmo314 committed:
Remove local GC allocation count flush in pystate.c

Removed flushing of thread's local GC allocation count before thread state deletion.

kevmo314 committed:
Merge branch 'main' into fix-gc-alloc-count-negative

colesbury reviewed (approved):
Thanks!
